### DISCUSSION TITLE
KdfMemory, KdfParallelism for argon2 configuration

### DISCUSSION THREAD

**Discussion by quexten** [dani-garcia/vaultwarden#3171] [Ideas]
Hi,

since https://github.com/bitwarden/server/pull/2583 has been merged now, this should be added to vaultwarden too, so that users can use and configure argon2.

Changes in short:
- Anywhere where kdfType/kdfIterations are sent, two additional **optional** parameters - kdfMemory, kdfParallelism - are sent 
- kdfType, kdfIterations are additionally stored in the User table in the database

**Comment by BlackDex** [comment]
We still need the GUI changes to be released. Without it, this will have no effect, and no way for use to fully test.

**Reply by tessus** [reply]
When the GUI changes are available, I'd like to give it a try and add it to the server.

**Reply by quexten** [reply]
@tessus The GUI pull request has been merged into master https://github.com/bitwarden/clients/commit/01091fe26081a0e59e517a245141227095724858

**Reply by tessus** [reply]
Thanks for the info. I am not sure if I can build the vaultwarden web-vault from master though. The last time I tried, it failed miserably.

But I can start adding argon to vw server. I should have time to look into this tomorrow.

**Reply by BlackDex** [reply]
I don't think adding Argon2 will benefit the server side in any way.
A 600_000 on the master-password is is more then enough.

**Reply by tessus** [reply]
Wait I thought I would have to change the server code to allow clients to use argon2. Or am I missing something? e.g. the 2 items that the OP mentioned at the beginning of the discussion.

I might have not expressed myself correctly. "adding argon to the server" was ambiguous, I give you that.

**Reply by BlackDex** [reply]
The only thing needed as far as i can tell it's adding a few new columns to the user table and some adjustments to the endpoints handling the user during pre-login and account creation.

**Reply by tessus** [reply]
Yep, I am going to have a go at it tommorow.

**Reply by quexten** [reply]
> The only thing needed as far as i can tell it's adding a few new columns to the user table and some adjustments to the endpoints handling the user during pre-login and account creation.

Yeah, correct. Server side does not run argon2. It only stores the extra parameters and adds some endpoints.

**Reply by quexten** [reply]
Web vault is now[ released.](https://github.com/bitwarden/clients/releases/tag/web-v2023.2.0)

**Reply by ghost** [reply]
> Web vault is now[ released.](https://github.com/bitwarden/clients/releases/tag/web-v2023.2.0)

client are also released on their git (already updated the client i maintain with vaultwarden rebrand), (only for self use not sure i can post it without breaching any licenses).

**Comment by pdarcos** [comment]
Chiming into the discussion here. I wonder if you guys have read [this analysis](https://palant.info/2023/01/23/bitwarden-design-flaw-server-side-iterations/) and this one too which is pretty critical https://dchest.com/2020/05/25/improving-storage-of-password-encrypted-secrets-in-end-to-end-encrypted-apps/

**Reply by BlackDex** [reply]
That is where all this is coming from. Also, nothing we can change on the Vaultwarden side to improve upon. And the password hash is already upgraded to 600_000 on the server side as default.

**Reply by quexten** [reply]
> Chiming into the discussion here. I wonder if you guys have read [this analysis](https://palant.info/2023/01/23/bitwarden-design-flaw-server-side-iterations/) and this one too which is pretty critical https://dchest.com/2020/05/25/improving-storage-of-password-encrypted-secrets-in-end-to-end-encrypted-apps/

Thanks for the notes, but I don't think they are not relevant here. This PR implements user kdf (argon2) parameter storage support in vaultwarden for compatibility with the official server. Protocol changes, as proposed in the articles you linked, should be suggested to the official project(s).

Edit: Ah didn't see BlackDex's comment while writing this. This was in response to the original comment.

**Reply by pdarcos** [reply]
Thanks for the quick reply @BlackDex 

As someone who has followed and tried to contribute to this project from the early days (when it was still called bitwarden-rs) I'm very impressed with the continuous improvements and quality of the code.

Well done guys!

But back to the topic, you don't think it is feasible to change the server code to address one of the main criticisms here? Or has this already been done?

They suggest:

"1) Instead of deriving one key (aka hash, aka verifier) with the password hashing function, derive two keys: serverEncKey and verifier. (The verifier has the same purpose as in the original scheme.) (Note: don’t use the password hashing function twice! Instead, with a modern password hashing function derive a 64-byte output and split it into two 32-byte keys, or if you’re stuck with PBKDF2, use HKDF to derive two different keys from a single 32-byte output of PBKDF2).

2) Encrypt protectedKey received from the client with serverEncKey and store the result of this encryption (serverProtectedKey) instead of protectedKey."

**Reply by quexten** [reply]
> They suggest:

> "1) Instead of deriving one key (aka hash, aka verifier) with the password hashing function, derive two keys: serverEncKey and verifier. (The verifier has the same purpose as in the original scheme.) (Note: don’t use the password hashing function twice! Instead, with a modern password hashing function derive a 64-byte output and split it into two 32-byte keys, or if you’re stuck with PBKDF2, use HKDF to derive two different keys from a single 32-byte output of PBKDF2).

>    Encrypt protectedKey received from the client with serverEncKey and store the result of this encryption (serverProtectedKey) instead of protectedKey."

Implementing this in vaultwarden would break compatibility with official clients. This would need to be implemented in the official server + clients first.

**Reply by ghost** [reply]
> > They suggest:
> 
> > "1) Instead of deriving one key (aka hash, aka verifier) with the password hashing function, derive two keys: serverEncKey and verifier. (The verifier has the same purpose as in the original scheme.) (Note: don’t use the password hashing function twice! Instead, with a modern password hashing function derive a 64-byte output and split it into two 32-byte keys, or if you’re stuck with PBKDF2, use HKDF to derive two different keys from a single 32-byte output of PBKDF2).
> 
> > Encrypt protectedKey received from the client with serverEncKey and store the result of this encryption (serverProtectedKey) instead of protectedKey."
> 
> Implementing this in vaultwarden would break compatibility with official clients. This would need to be implemented in the official server + clients first.

they have updated it and i have updated the client manually, so the test can be done now i think :), it mean that next version of vaultwarden will have 2023.2.0 instead of 2023.1.0 ?
